File: examples/python/liskov-substitution-principle/bad.py
import sys
import os

sys.path.append(os.path.abspath(os.path.join(
    os.path.dirname(__file__), '..')
))

import math
import json
from models.ShapeInterface import ShapeInterface
from models.Circle import Circle
from models.Square import Square

class AreaCalculator:
    def __init__(self, shapes: list):
        self.shapes = shapes

    def sum(self):

        sum_area = 0

        for shape in self.shapes:
            if(issubclass(type(shape), ShapeInterface)):
                sum_area += shape.area()
            else:
                raise SystemError
        return sum_area


class SumCalculatorOutputter:
    def __init__(self, calculator: AreaCalculator):
        self.calculator = calculator
    
    def JSON(self):
        return json.dumps({'sum': self.calculator.sum()}, sort_keys=True)        

    def Message(self) -> str:
        return f"Sum of the areas of provided shapes: {self.calculator.sum():.2f}"

class VolumeCalculator(AreaCalculator):
    def __init__(self, shapes: list):
        super().__init__(shapes)
    
    def sum(self) -> float:
        vol_sum = 0
        for shape in self.shapes:
            if(issubclass(type(shape), ShapeInterface)):
                # think about shape return volume 
                vol_sum += 0
            else:
                raise SystemError

        return vol_sum
        

'''
Here is an example with a collection of three shapes:

a circle with a radius of 2
a square with a length of 5
a second square with a length of 6
'''

shapes = [
  Circle(2),
  Square(5),
  Square(6),
]

areas = VolumeCalculator(shapes)
output = SumCalculatorOutputter(areas)

print(output.JSON())
print(output.Message())

File: examples/csharp/OpenClosedPrinciple/Bad.cs
using System;
using System.Text.Json;

namespace OCP.Bad
{
    public class Circle
    {
        public int Radius { get; private set; }

        public Circle(int radius)
        {
            this.Radius = radius;
        }
    }

    public class Square
    {
        public int Length { get; private set; }

        public Square(int length)
        {
            this.Length = length;
        }
    }

    public class AreaCalculator
    {
        private List<object> shapes = null;
        public AreaCalculator(List<object> shapes)
        {
            this.shapes = shapes;
        }

        public double Sum()
        {
            double sumArea = 0.0;

            if (this.shapes == null)
                return 0;
            
            foreach (var shape in this.shapes)
            {
                if (shape.GetType() == typeof(Square))
                    sumArea += Math.Pow(((Square)shape).Length, 2);
                if (shape.GetType() == typeof(Circle))
                    sumArea += Math.PI * Math.Pow(((Circle)shape).Radius, 2);
            }
            
            return sumArea;
        }
    }

    public class SumCalculatorOutputter
    {
        private AreaCalculator calculator;

        public SumCalculatorOutputter(AreaCalculator calculator)
        {
            this.calculator = calculator;
        }

        public string JSON()
        {
            // create object
            var response = new {
                sum = this.calculator.Sum()
            };

            return JsonSerializer.Serialize(response);
        }

        public string Message()
        {
            return String.Format("Sum of the areas of provided shapes: {0:0.00}", 
                this.calculator.Sum());

        }
    }
}
